project/grammar_tools.py

```diff
@@ -337,23 +366,286 @@ def is_wcnf(acnf: CFG, cfg: CFG) -> bool:
+
+            head = Variable(head_text.strip())
+            if head in variables:
+                raise ValueError("Only one production for each variable is required")
+            variables.add(head)
+
+            body = Regex(body_text.strip())
+
+            productions.add(ECFGProduction(head, body))
+
+        return cls(
+            variables=variables, start_symbol=start_symbol, productions=productions
         )
-    )
 
-    cnf_epsilon_productions = set(filter(lambda prod: not prod.body, cnf_productions))
+    @classmethod
+    def from_cfg(cls, cfg: CFG) -> "ECFG":
+        """
+        Get an Extended Context Free Grammar from Context Free Grammar.
+
+        Parameters
+        ----------
+        cfg: CFG
+            CFG to convert
+
+        Returns
+        -------
+        ECFG:
+            ECFG equivalent to given CFG
+        """
+
+        productions = dict()
+
+        for cfg_production in cfg.productions:
+            body = Regex(
+                " ".join(body_object.value for body_object in cfg_production.body)
+                if cfg_production.body
+                else "epsilon"
+            )
+
+            if cfg_production.head not in productions:
+                productions[cfg_production.head] = body
+            else:
+                productions[cfg_production.head] = productions.get(
+                    cfg_production.head
+                ).union(body)
+
+        ecfg_productions = (
+            ECFGProduction(head, body) for head, body in productions.items()
+        )
 
-    for epsilon_production in cfg_epsilon_productions:
-        if epsilon_production not in cnf_epsilon_productions:
-            return False
+        return cls(
+            variables=cfg.variables,
+            start_symbol=cfg.start_symbol,
+            productions=ecfg_productions,
+        )
 
-    return True
+
+def get_ecfg_from_cfg(cfg: CFG) -> ECFG:
+    """
+    Get an Extended Context Free Grammar from Context Free Grammar.
+
+    Parameters
+    ----------
+    cfg: CFG
+        CFG to convert
+
+    Returns
+    -------
+    ECFG:
+        ECFG equivalent to given CFG
+    """
+
+    return ECFG.from_cfg(cfg)
+
+
+def cyk(word: str, cfg: CFG) -> bool:
+    """
+    Checks whether grammar derive the word.
+
+    This function is applicable to any CFG.
+
+    Parameters
+    ----------
+    cfg: CFG
+        A CFG to derive a word
+    word:
+        A word to derive in cfg
+
+    Returns
+    -------
+    bool:
+        Whether grammar derive the word
+    """
+
+    word_len = len(word)
+
+    if not word_len:
+        return cfg.generate_epsilon()
+
+    cnf = cfg.to_normal_form()
+
+    terminal_productions = [
+        production for production in cnf.productions if len(production.body) == 1
+    ]
+    variable_productions = [
+        production for production in cnf.productions if len(production.body) == 2
+    ]
+
+    matrix = [[set() for _ in range(word_len)]] * word_len
```

Comment:
```python
>>> d = [[set() for _ in range(2)]] * 2
>>> d
[[set(), set()], [set(), set()]]
>>> d[0][0].add(1)
>>> d
[[{1}, set()], [{1}, set()]]
```
Я думаю, Вы такого не ожидали.

Лучше сделать так.
```suggestion
    matrix = [[set() for _ in range(word_len)] for _ in range(word_len)]
```

Тогда будет
```python
>>> d = [[set() for _ in range(2)] for _ in range(2)] 
>>> d
[[set(), set()], [set(), set()]]
>>> d[0][0].add(1)
>>> d
[[{1}, set()], [set(), set()]]
```